src/koopman.jl

```diff
@@ -7,13 +7,88 @@ struct MonteCarlo <: AbstractExpectationAlgorithm end
 @inline tuplejoin(x, y) = (x..., y...)
 @inline tuplejoin(x, y, z...) = (x..., tuplejoin(y, z...)...)
 
-_rand(x::T) where T <: Sampleable = rand(x)
-_rand(x) = x
-
 function __make_map(prob::ODEProblem, args...; kwargs...)
     (u,p) -> solve(remake(prob,u0=u,p=p), args...; kwargs...)
 end
 
+"""
+    solve(prob, expalg::Koopman, args; kwargs...)
+
+Solves the ExpectationProblem using via the Koopman expectation
+
+Both args and kwargs are passed to DifferentialEquation solver
+
+Special kwargs:
+    - maxiters: max quadrature iterations (default 1000000)
+    - batch: solve quadrature using n-batch processing (default 0, off)
+    - quadalg: quadrature algorithm to use (default HCubatureJL())
+    - ireltol, iabstol: integration relative and absolute tolerances (default 1e-2, 1e-2)
+"""
+function DiffEqBase.solve(prob::ExpectationProblem, expalg::Koopman, args...;
+    maxiters=1000000,
+    batch=0,
+    quadalg=HCubatureJL(),
+    ireltol=1e-2, iabstol=1e-2,
+    kwargs...)
+
+    jargs = tuplejoin(args)
+    jkwargs = tuplejoin(prob.kwargs, kwargs)
+
+    # build the integrand for ∫(Ug)(x) * f(x) dx 
+    integrand = function(fq, xq, pq)
+        # convert to physical x and p
+        (_x, _p) = prob.comp_func(prob.to_phys(xq,pq)...)
+
+        # solve ODE and evaluate observable
+        if batch == 0
+            # scalar solution
+            _prob = remake(prob.ode_prob, u0=_x, p=_p)
+            Ug = prob.g(solve(_prob, jargs...; jkwargs...))
+            f0 = prob.f0_func(_x, _p)
+        else
+            # ensemble solution
+            prob_func(prob, i, repeat) = remake(prob, u0=_x[:,i], p=_p[:,i])
+            output_func(sol,i) = prob.g(sol), false
+            _prob = EnsembleProblem(prob.ode_prob, prob_func=prob_func, output_func=output_func)
+            Ug = solve(_prob, jargs...; trajectories=size(xq,2), jkwargs...)[:]
+            f0 = map(prob.f0_func, zip(_x,_p))
+        end
+
+        fq .= Ug .* f0
```

Comment:
@ChrisRackauckas this causes `HCubatureJL` to fail for `nout=1`...if I switch the integrand to oop, it passes....

I dug into it, and it executes the exact same quadrature evaluation points and returning integrand values for both directly returning a scalar and in-place for a `length=1` array. The quadrature alg seems to keep evaluating more points for the latter therefore returning a different value.

Any ideas?
FYI @agerlach

Reply:
HCubature will internally call different algorithms depending on if scalar bounds or array. This may explain it??? See hcubature vs hquadrature https://github.com/JuliaMath/HCubature.jl